core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1050,7 +1055,11 @@ class Log(@volatile var dir: File,
         deletable.foreach(deleteSegment)
         logStartOffset = math.max(logStartOffset, segments.firstEntry().getValue.baseOffset)
         leaderEpochCache.clearAndFlushEarliest(logStartOffset)
-        producerStateManager.evictUnretainedProducers(logStartOffset)
+
+        // Update the producer state with the new log start offset, which we cause any non-retained producers to
```

Comment:
Hmm.. In `truncateAndReload`, we first delete any snapshots which are out of range. Is there a case where the start and end offsets passed to `truncateAndReload` would be out of sync with the log segments?